plutus-core-interpreter/src/Language/PlutusCore/Interpreter/CekMachine.hs

```diff
@@ -198,23 +199,33 @@ applyStagedBuiltinName (StaticStagedBuiltinName  name) args =
 
 -- | Evaluate a term in an environment using the CEK machine.
 evaluateCekCatchIn
-    :: CekEnv -> Plain Term -> Either CekMachineException EvaluationResult
+    :: CekEnv -> Plain Term -> Either CekMachineException EvaluationResultDef
 evaluateCekCatchIn cekEnv = runCekM cekEnv . computeCek []
 
 -- | Evaluate a term using the CEK machine.
 evaluateCekCatch
-    :: DynamicBuiltinNameMeanings -> Plain Term -> Either CekMachineException EvaluationResult
+    :: DynamicBuiltinNameMeanings -> Plain Term -> Either CekMachineException EvaluationResultDef
 evaluateCekCatch means = evaluateCekCatchIn $ CekEnv means mempty
 
 -- | Evaluate a term using the CEK machine. May throw a 'CekMachineException'.
-evaluateCek :: DynamicBuiltinNameMeanings -> Term TyName Name () -> EvaluationResult
+evaluateCek :: DynamicBuiltinNameMeanings -> Term TyName Name () -> EvaluationResultDef
 evaluateCek = either throw id .* evaluateCekCatch
 
+-- The implementation is a bit of a hack.
 readDynamicBuiltinCek
-    :: KnownDynamicBuiltinType dyn => Term TyName Name () -> Either CekMachineException (Maybe dyn)
-readDynamicBuiltinCek = readDynamicBuiltin evaluateCekCatch
+    :: KnownDynamicBuiltinType dyn
+    => Term TyName Name () -> Either CekMachineException (Maybe dyn)
+readDynamicBuiltinCek term = do
+    res <- readDynamicBuiltin evaluateCekCatch term
+    case runExceptT res of
+        EvaluationFailure        -> Right Nothing
+        EvaluationSuccess errOrX -> bimap unreadableBuiltin Just errOrX where
```

Comment:
I think this is quite a bit less readable than a pattern match.

Reply:
Agreed.